src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,242 @@
+            return first.element;
+        }
+        if (index == size - 1) {
+            return last.element;
+        }
+        if (index <= size / 2) {
+            int count = 0;
+            Node<T> node = first;
+            while (count != index) {
+                node = node.next;
+                count++;
+            }
+            return node.element;
+        } else {
+            int count = size - 1;
+            Node<T> node = last;
+            while (count != index) {
+                node = node.prev;
+                count--;
+            }
+            return node.element;
+        }
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException();
+        }
+        Node<T> currentNode;
+        if (index <= size / 2) {
+            int count = 0;
+            currentNode = first;
+            while (count != index) {
+                currentNode = currentNode.next;
+                count++;
+            }
+        } else {
+            int count = size - 1;
+            currentNode = last;
+            while (count != index) {
+                currentNode = currentNode.prev;
+                count--;
+            }
+        }
+        T oldValue = currentNode.element;
+        currentNode.element = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException();
+        }
+        if (index == 0 && size == 1) {
+            T element = first.element;
+            size--;
+            return element;
+        }
+        if (index == 0 && size > 1) {
+            T element = first.element;
+            first = first.next;
+            size--;
+            return element;
+        }
+        if (index == size - 1) {
+            last.prev.next = null;
+            last = last.prev;
+            T t = last.element;
+            size--;
+            return t;
+        }
+        Node<T> currentNode;
+        if (index <= size / 2) {
+            int count = 0;
+            currentNode = first;
+            while (count != index) {
+                currentNode = currentNode.next;
+                count++;
+            }
+        } else {
+            int count = size - 1;
+            currentNode = last;
+            while (count != index) {
+                currentNode = currentNode.prev;
+                count--;
+            }
+        }
+        currentNode.prev.next = currentNode.next;
+        currentNode.next.prev = currentNode.prev;
+        size--;
+        return currentNode.element;
     }
 
     @Override
     public boolean remove(T object) {
+        Node<T> currentNode = first;
+        while (currentNode != null) {
+            if (object == null) {
+                if (currentNode.element == null) {
+                    removeNode(currentNode);
```

Comment:
Create a separate method `unlink(Node node)` to handle the logic of unlinking a node. This logic will be used in 2 methods: when we remove from the list by index and by value.

Reply:
Done